File: impl/java/src/test/java/sh/bubblefish/npamp/CertVerifyKat.java
// Standards-derived, NON-CIRCULAR known-answer test for the draft-00 CertVerify
// (binding spec/10 section 6.1; RFC 8446 section 4.4.3 structure; Ed25519 per RFC 8032).
// The value is u16(0x0807) || Ed25519(priv, signing_input), signing_input =
// 64*0x20 || context || 0x00 || TH. Java mirror of the Go/TS/Python reference tests
// against the SAME pinned vector (test-vectors/v1/certverify-kat.json).
//
// Three legs: ANCHOR (the src Ed25519 helpers reproduce RFC 8032 TEST1/TEST2 keys +
// signatures), ORACLE (rebuild signing_input by hand + sign with an independently
// constructed key, no src signing functions), IMPL (certVerifySigningInput +
// signCertVerify reproduce the vector; verifyCertVerify accepts the correct value but
// rejects a role/context mismatch, a wrong transcript, a wrong scheme, and a truncated
// signature). Run via:
//   javac -d <out> src/main/java/sh/bubblefish/npamp/*.java src/test/java/sh/bubblefish/npamp/*.java
//   java  -cp <out> sh.bubblefish.npamp.CertVerifyKat
package sh.bubblefish.npamp;

import static sh.bubblefish.npamp.Kat.fromHex;
import static sh.bubblefish.npamp.Kat.sat;
import static sh.bubblefish.npamp.Kat.toHex;

import java.nio.charset.StandardCharsets;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.Signature;

public final class CertVerifyKat {

    private CertVerifyKat() {
    }

    static final String CERTVERIFY_KAT_SHA256 = "19afd438c3036fd7d51481e5e6e91cc73010d76cb94aa2082c7752c8ba714d3f";

    private static int failures = 0;

    private static void check(String name, boolean ok, String detail) {
        if (ok) {
            System.out.println("ok   - " + name);
        } else {
            System.out.println("FAIL - " + name + (detail.isEmpty() ? "" : ": " + detail));
            failures++;
        }
    }

    /** Oracle signing input, built by hand independently of certVerifySigningInput. */
    private static byte[] oracleSigningInput(String ctx, byte[] th) {
        byte[] c = ctx.getBytes(StandardCharsets.US_ASCII);
        byte[] out = new byte[64 + c.length + 1 + th.length];
        int p = 0;
        for (int i = 0; i < 64; i++) {
            out[p++] = 0x20;
        }
        System.arraycopy(c, 0, out, p, c.length);
        p += c.length;
        out[p++] = 0x00;
        System.arraycopy(th, 0, out, p, th.length);
        return out;
    }

    private static byte[] oracleSign(PrivateKey priv, byte[] msg) {
        try {
            Signature s = Signature.getInstance("Ed25519");
            s.initSign(priv);
            s.update(msg);
            return s.sign();
        } catch (Exception e) {
            throw new RuntimeException("oracle ed25519 sign failed", e);
        }
    }

    /** ANCHOR: the src Ed25519 helpers reproduce RFC 8032 TEST1/TEST2 pubkeys + signatures. */
    private static void anchor(Object root) {
        for (String tc : new String[]{"test1", "test2"}) {
            byte[] seed = fromHex(sat(root, "rfc8032_ed25519", tc, "seed"));
            byte[] msg = fromHex(sat(root, "rfc8032_ed25519", tc, "message"));
            String wantPub = sat(root, "rfc8032_ed25519", tc, "public_key");
            String wantSig = sat(root, "rfc8032_ed25519", tc, "signature");

            PrivateKey priv = Handshake.ed25519PrivateKeyFromSeed(seed);
            byte[] sig = oracleSign(priv, msg);
            check("anchor: RFC 8032 " + tc + " signature", toHex(sig).equals(wantSig),
                    "got " + toHex(sig) + " want " + wantSig);

            // ed25519PublicKeyFromRaw decodes the published raw pubkey and verifies the
            // published signature, proving the raw-point decoder round-trips (RFC 8032).
            PublicKey pub = Handshake.ed25519PublicKeyFromRaw(fromHex(wantPub));
            boolean verified;
            try {
                Signature v = Signature.getInstance("Ed25519");
                v.initVerify(pub);
                v.update(msg);
                verified = v.verify(fromHex(wantSig));
            } catch (Exception e) {
                verified = false;
            }
            check("anchor: RFC 8032 " + tc + " pubkey-from-raw verifies", verified, "");
        }
    }

    /** ORACLE: rebuild signing_input by hand + sign with an independent key (guards the vector). */
    private static void oracle(Object root) {
        String[][] cases = {
                {"server", "server", "server_seed", "th_sid", "signing_input_server", "signature_server"},
                {"client", "client", "client_seed", "th_cid", "signing_input_client", "signature_client"},
        };
        for (String[] c : cases) {
            String ctx = sat(root, "contexts", c[1]);
            byte[] seed = fromHex(sat(root, "npamp_inputs", c[2]));
            byte[] th = fromHex(sat(root, "npamp_inputs", c[3]));
            byte[] si = oracleSigningInput(ctx, th);
            check("oracle: " + c[0] + " signing_input", toHex(si).equals(sat(root, "expected", c[4])),
                    "got " + toHex(si));
            byte[] sig = oracleSign(Handshake.ed25519PrivateKeyFromSeed(seed), si);
            check("oracle: " + c[0] + " signature", toHex(sig).equals(sat(root, "expected", c[5])),
                    "got " + toHex(sig));
        }
    }

    /** IMPL: certVerifySigningInput + signCertVerify reproduce the vector; verifyCertVerify guards. */
    private static void impl(Object root) {
        if (!Handshake.CONTEXT_SERVER_CERTVERIFY.equals(sat(root, "contexts", "server"))
                || !Handshake.CONTEXT_CLIENT_CERTVERIFY.equals(sat(root, "contexts", "client"))) {
            check("impl: context constants match spec section 6.1", false,
                    "server=" + Handshake.CONTEXT_SERVER_CERTVERIFY + " client=" + Handshake.CONTEXT_CLIENT_CERTVERIFY);
            return;
        }
        Object[][] cases = {
                {"server", Boolean.TRUE, "server_seed", "server_pub", "th_sid", "signing_input_server", "certverify_value_server"},
                {"client", Boolean.FALSE, "client_seed", "client_pub", "th_cid", "signing_input_client", "certverify_value_client"},
        };
        for (Object[] c : cases) {
            String name = (String) c[0];
            boolean isServer = (Boolean) c[1];
            PrivateKey priv = Handshake.ed25519PrivateKeyFromSeed(fromHex(sat(root, "npamp_inputs", (String) c[2])));
            PublicKey pub = Handshake.ed25519PublicKeyFromRaw(fromHex(sat(root, "npamp_inputs", (String) c[3])));
            byte[] th = fromHex(sat(root, "npamp_inputs", (String) c[4]));

            String gotSI = toHex(Handshake.certVerifySigningInput(isServer, th));
            check("impl: " + name + " certVerifySigningInput", gotSI.equals(sat(root, "expected", (String) c[5])),
                    "got " + gotSI);

            byte[] val = Handshake.signCertVerify(priv, isServer, th);
            check("impl: " + name + " signCertVerify value", toHex(val).equals(sat(root, "expected", (String) c[6])),
                    "got " + toHex(val));

            check("impl: " + name + " verifyCertVerify accepts",
                    Handshake.verifyCertVerify(pub, isServer, th, val), "");

            // Domain separation: the opposite role must FAIL (different context string).
            check("impl: " + name + " rejects role/context mismatch",
                    !Handshake.verifyCertVerify(pub, !isServer, th, val), "");

            // Transcript binding: a different transcript hash must FAIL.
            byte[] wrongTh = th.clone();
            wrongTh[0] ^= 0x01;
            check("impl: " + name + " rejects wrong transcript",
                    !Handshake.verifyCertVerify(pub, isServer, wrongTh, val), "");

            // Scheme guard: a non-Ed25519 scheme code point must FAIL.
            byte[] badScheme = val.clone();
            badScheme[0] = (byte) ((Npamp.SIG_MLDSA87 >>> 8) & 0xFF);
            badScheme[1] = (byte) (Npamp.SIG_MLDSA87 & 0xFF);
            check("impl: " + name + " rejects non-Ed25519 scheme",
                    !Handshake.verifyCertVerify(pub, isServer, th, badScheme), "");

            // Length guard: an Ed25519 signature is exactly 64 octets; a truncated value must FAIL.
            byte[] truncated = new byte[val.length - 1];
            System.arraycopy(val, 0, truncated, 0, truncated.length);
            check("impl: " + name + " rejects truncated signature",
                    !Handshake.verifyCertVerify(pub, isServer, th, truncated), "");
        }
    }

    public static void main(String[] args) throws Exception {
        Object root = Kat.loadPinned(args, "certverify-kat.json", CERTVERIFY_KAT_SHA256);
        anchor(root);
        oracle(root);
        impl(root);
        System.out.println(failures == 0 ? "ALL PASS (certverify KAT: anchor+oracle+impl)" : ("FAILURES: " + failures));
        System.exit(failures == 0 ? 0 : 1);
    }
}

File: impl/kotlin/src/test/kotlin/sh/bubblefish/npamp/CertVerifyKat.kt
// Standards-derived, NON-CIRCULAR known-answer test for the draft-00 CertVerify
// (binding spec/10 §6.1; RFC 8446 §4.4.3 structure; Ed25519 per RFC 8032). The value
// is u16(0x0807) || Ed25519(priv, signing_input), signing_input = 64*0x20 || context ||
// 0x00 || TH. Kotlin mirror of the Go/TS/Python reference tests against the SAME pinned
// vector (test-vectors/v1/certverify-kat.json).
//
// Three legs: ANCHOR (the src Ed25519 helpers reproduce RFC 8032 TEST1/TEST2 signature +
// public-key decode), ORACLE (rebuild signing_input by hand + sign with an independently
// constructed JDK key, no src signing functions), IMPL (certVerifySigningInput +
// signCertVerify reproduce the vector; verifyCertVerify accepts the correct value but
// rejects a role/context mismatch, a wrong transcript, a wrong scheme, and a truncated sig).
package sh.bubblefish.npamp

import java.io.ByteArrayOutputStream
import java.nio.charset.StandardCharsets
import java.security.KeyFactory
import java.security.PrivateKey
import java.security.Signature
import java.security.spec.EdECPrivateKeySpec
import java.security.spec.NamedParameterSpec
import kotlin.system.exitProcess

object CertVerifyKat {

    private const val PIN = "19afd438c3036fd7d51481e5e6e91cc73010d76cb94aa2082c7752c8ba714d3f"

    private var failures = 0

    private fun leg(name: String, fn: () -> Unit) {
        try {
            fn()
            println("PASS $name")
        } catch (e: Throwable) {
            println("FAIL $name: ${e.message}")
            failures++
        }
    }

    /** Builds an Ed25519 private key from its seed, inline and independent of Handshake. */
    private fun oraclePriv(seed: ByteArray): PrivateKey =
        KeyFactory.getInstance("Ed25519").generatePrivate(EdECPrivateKeySpec(NamedParameterSpec.ED25519, seed))

    /** Raw Ed25519 sign over [msg], independent of signCertVerify. */
    private fun oracleSign(priv: PrivateKey, msg: ByteArray): ByteArray {
        val s = Signature.getInstance("Ed25519")
        s.initSign(priv)
        s.update(msg)
        return s.sign()
    }

    /** The §6.1 signing input, built by hand and independent of certVerifySigningInput. */
    private fun oracleSigningInput(ctx: String, th: ByteArray): ByteArray {
        val out = ByteArrayOutputStream()
        for (i in 0 until 64) {
            out.write(0x20)
        }
        val c = ctx.toByteArray(StandardCharsets.US_ASCII)
        out.write(c, 0, c.size)
        out.write(0x00)
        out.write(th, 0, th.size)
        return out.toByteArray()
    }

    @JvmStatic
    fun main(args: Array<String>) {
        val k = KatSupport.loadPinned(args, "certverify-kat.json", PIN)
        val nn = KatSupport.obj(k, "npamp_inputs")
        val e = KatSupport.obj(k, "expected")
        val c = KatSupport.obj(k, "contexts")

        leg("certverify_kat_rfc8032_anchor") {
            val rfc = KatSupport.obj(k, "rfc8032_ed25519")
            for (label in listOf("test1", "test2")) {
                val v = KatSupport.obj(rfc, label)
                val seed = KatSupport.fromHex(v["seed"] as String)
                val msg = KatSupport.fromHex(v["message"] as String)
                // src signer reproduces the RFC 8032 signature.
                val priv = Handshake.ed25519PrivateKeyFromSeed(seed)
                val gotSig = KatSupport.toHex(oracleSign(priv, msg))
                check(gotSig == v["signature"]) { "$label signature != RFC 8032\n  got  $gotSig\n  want ${v["signature"]}" }
                // src public-key decoder reproduces a key that verifies the RFC 8032 signature.
                val pub = Handshake.ed25519PublicKeyFromRaw(KatSupport.fromHex(v["public_key"] as String))
                val verifier = Signature.getInstance("Ed25519")
                verifier.initVerify(pub)
                verifier.update(msg)
                check(verifier.verify(KatSupport.fromHex(v["signature"] as String))) {
                    "$label: ed25519PublicKeyFromRaw produced a key that rejected the RFC 8032 signature"
                }
            }
        }

        leg("certverify_kat_oracle") {
            data class Case(val name: String, val ctx: String, val seed: String, val th: String, val wantSI: String, val wantSig: String)
            val cases = listOf(
                Case("server", c["server"] as String, nn["server_seed"] as String, nn["th_sid"] as String, e["signing_input_server"] as String, e["signature_server"] as String),
                Case("client", c["client"] as String, nn["client_seed"] as String, nn["th_cid"] as String, e["signing_input_client"] as String, e["signature_client"] as String),
            )
            for (cs in cases) {
                val si = oracleSigningInput(cs.ctx, KatSupport.fromHex(cs.th))
                check(KatSupport.toHex(si) == cs.wantSI) { "[${cs.name}] oracle signing_input != vector" }
                val gotSig = KatSupport.toHex(oracleSign(oraclePriv(KatSupport.fromHex(cs.seed)), si))
                check(gotSig == cs.wantSig) { "[${cs.name}] oracle signature != vector\n  got  $gotSig\n  want ${cs.wantSig}" }
            }
        }

        leg("certverify_kat_impl") {
            check(Handshake.CONTEXT_SERVER_CERTVERIFY == c["server"]) { "server context constant drifted from spec §6.1" }
            check(Handshake.CONTEXT_CLIENT_CERTVERIFY == c["client"]) { "client context constant drifted from spec §6.1" }
            data class Case(val name: String, val isServer: Boolean, val seed: String, val pub: String, val th: String, val wantSI: String, val wantVal: String)
            val cases = listOf(
                Case("server", true, nn["server_seed"] as String, nn["server_pub"] as String, nn["th_sid"] as String, e["signing_input_server"] as String, e["certverify_value_server"] as String),
                Case("client", false, nn["client_seed"] as String, nn["client_pub"] as String, nn["th_cid"] as String, e["signing_input_client"] as String, e["certverify_value_client"] as String),
            )
            for (cs in cases) {
                val priv = Handshake.ed25519PrivateKeyFromSeed(KatSupport.fromHex(cs.seed))
                val pub = Handshake.ed25519PublicKeyFromRaw(KatSupport.fromHex(cs.pub))
                val th = KatSupport.fromHex(cs.th)

                val gotSI = KatSupport.toHex(Handshake.certVerifySigningInput(cs.isServer, th))
                check(gotSI == cs.wantSI) { "[${cs.name}] certVerifySigningInput != vector\n  got  $gotSI\n  want ${cs.wantSI}" }

                val value = Handshake.signCertVerify(priv, cs.isServer, th)
                val gotVal = KatSupport.toHex(value)
                check(gotVal == cs.wantVal) { "[${cs.name}] signCertVerify value != vector\n  got  $gotVal\n  want ${cs.wantVal}" }

                check(Handshake.verifyCertVerify(pub, cs.isServer, th, value)) { "[${cs.name}] verifyCertVerify rejected the correct value" }
                // Domain separation: the opposite role must FAIL (different context string).
                check(!Handshake.verifyCertVerify(pub, !cs.isServer, th, value)) { "[${cs.name}] accepted a role/context mismatch" }
                // Transcript binding: a different transcript hash must FAIL.
                val wrong = th.copyOf()
                wrong[0] = (wrong[0].toInt() xor 0x01).toByte()
                check(!Handshake.verifyCertVerify(pub, cs.isServer, wrong, value)) { "[${cs.name}] accepted a wrong transcript hash" }
                // Scheme guard: a non-Ed25519 scheme code point must FAIL.
                val badScheme = value.copyOf()
                badScheme[0] = ((Npamp.SIG_MLDSA87 ushr 8) and 0xFF).toByte()
                badScheme[1] = (Npamp.SIG_MLDSA87 and 0xFF).toByte()
                check(!Handshake.verifyCertVerify(pub, cs.isServer, th, badScheme)) { "[${cs.name}] accepted a non-Ed25519 scheme" }
                // Length guard: an Ed25519 signature is exactly 64 octets; a truncated value must FAIL.
                check(!Handshake.verifyCertVerify(pub, cs.isServer, th, value.copyOf(value.size - 1))) { "[${cs.name}] accepted a truncated signature" }
            }
        }

        println(if (failures == 0) "ALL PASS (3/3)" else "FAILURES: $failures")
        exitProcess(if (failures == 0) 0 else 1)
    }
}
